x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/StepInfoExceptionWrapper.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.indexlifecycle;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Wraps an Exception so that it can be passed to things which require an
+ * {@link org.elasticsearch.common.xcontent.ToXContentObject}.
+ */
+public class StepInfoExceptionWrapper implements ToXContentObject {
+    private final Throwable exception;
+
+    public StepInfoExceptionWrapper(Throwable exception) {
+        this.exception = Objects.requireNonNull(exception);
+    }
+
+    Class<? extends Throwable> getExceptionType() {
+        return exception.getClass();
+    }
+
+    String getMessage() {
+        return exception.getMessage();
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        ElasticsearchException.generateThrowableXContent(builder, params, exception);
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        StepInfoExceptionWrapper that = (StepInfoExceptionWrapper) o;
+        return Objects.equals(exception.getMessage(), that.exception.getMessage())
+            && Objects.equals(exception.getClass(), that.exception.getClass());
```

Comment:
Is there a reason we shouldn't do `Objects.equals(exception, that.exception)`? Do we want to have exceptions with different stacktraces be considered the same?

Reply:
I went back and forth on this too, but there's two things:
1) This is for serializing to `step_info`, and only the exception type and message get into the serialized version, so you can argue that it makes sense to check those here.
2) If just check equality of the exception here, then do you know of a good way to get [this test](https://github.com/elastic/elasticsearch/pull/34206/files#diff-4275418336a01afcff7e382c78312c4fR363) working?

I'd be delighted to go with a different approach here, I don't really like this class if I'm being honest.  Do you have a recommendation for a different way to approach this?

Reply:
One option might be for this class to hold the already serialised form of the exception, but I'm not sure if that is better or worse than the current solution